src/app/InteractionModelEngine.cpp

```diff
@@ -1408,5 +1408,32 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
     return numDirtySubscriptions;
 }
 
+void InteractionModelEngine::OnFabricRemoved(FabricIndex aFabricIndex)
+{
+    mReadHandlers.ForEachActiveObject([this, aFabricIndex](ReadHandler * handler) {
+        if (handler->GetAccessingFabricIndex() == aFabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete read/subscription handler with FabricIndex: %u",
+                            handler->GetAccessingFabricIndex());
+            mReadHandlers.ReleaseObject(handler);
+        }
+
+        return Loop::Continue;
+    });
+
+    for (auto & handler : mWriteHandlers)
+    {
+        if (!(handler.IsFree()) && handler.GetAccessingFabricIndex() == aFabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete write handler with FabricIndex: %u",
+                            handler.GetAccessingFabricIndex());
+            handler.Close();
+        }
+    }
+
+    // App code may hold the reference for Async Command Handler, when specific fabric index is removed, if we are not cleaning up
+    // the corresponding CommandHandler holed by the app, the app may fail to send Command Response using invalid exchange, then it
+    // is still able to close
```

Comment:
```suggestion
    // Applications may hold references to CommandHandler instances for async command processing.
    // Therefore we can't forcible destroy CommandHandlers here.  Their exchanges will get closed by
    // the fabric removal, though, so they will fail when they try to actually send their command response
    // and will close at that point.
```
As a followup, maybe CommandHandler should track its handles so it's able to release them all or something....  We'd need to check consumers to make sure they do the requisite null-checks.

Reply:
https://github.com/project-chip/connectedhomeip/issues/21739